copy store through a temporary file

domsj opened the pull request:
Still waiting for CI, but feel free to take a look already.

Tested by manually removing arakoon_0.db from a node and then starting it.
Resulted in the following output:
Jul 19 15:34:58 0928: (main|debug): stat /tmp/arakoon/arakoon_0/head.db
Jul 19 15:34:58 0930: (main|debug): stat /tmp/arakoon/arakoon_0/arakoon_0.db
Jul 19 15:34:58 0931: (main|debug): stat /tmp/arakoon/arakoon_0/arakoon_0.db.tmp
Jul 19 15:34:58 0931: (main|debug): copy_file /tmp/arakoon/arakoon_0/head.db /tmp/arakoon/arakoon_0/arakoon_0.db.tmp
Jul 19 15:34:58 1110: (main|debug): done: copy_file
Jul 19 15:34:58 1111: (main|debug): rename /tmp/arakoon/arakoon_0/arakoon_0.db.tmp -> /tmp/arakoon/arakoon_0/arakoon_0.db

domsj commented on src/node/local_store.ml:
```diff
@@ -79,7 +79,26 @@ let copy_store2 old_location new_location overwrite =
+          begin
+            if dest_exists
+            then
+              File_system.unlink new_location
```
Hmm, this probably isn't needed and is handled by the rename below too right?
(Otoh, I don't think it hurts either?)

NicolasT replied:
You can't rename to an existing location.

Next to that, unlink might fail if for some reason the provided path is removed already by something else, so might want some error detection.

NicolasT commented on src/node/local_store.ml:
```diff
@@ -79,7 +79,26 @@ let copy_store2 old_location new_location overwrite =
+          begin
+            if tmp_exists
+            then
+              File_system.unlink tmp_file
```
Same as above

NicolasT commented on src/node/local_store.ml:
```diff
@@ -79,7 +79,26 @@ let copy_store2 old_location new_location overwrite =
+            else
+              Lwt.return ()
+          end >>= fun () ->
+          File_system.copy_file old_location tmp_file >>= fun () ->
```
`copy_file` doesn't fsync.

NicolasT commented on src/node/local_store.ml:
```diff
@@ -79,7 +79,26 @@ let copy_store2 old_location new_location overwrite =
+              Lwt.return ()
+          end >>= fun () ->
+          File_system.copy_file old_location tmp_file >>= fun () ->
+          File_system.rename tmp_file new_location
```
Need to `fsync` dirents?

domsj committed:
straighten catchup interface
correct i and vo to start election with when starting a node

domsj committed:
ensure the previous value being pushed really is the previous

domsj committed:
ensure a MasterSet is pushed after reaching consensus on another value when running for master

domsj committed:
more fine grained ElectionTimeout also takes i into account

domsj committed:
when discovering a new master take their n

domsj committed:
Merge remote-tracking branch 'refs/remotes/origin/election-timeout' into 1.7

domsj committed:
Merge remote-tracking branch 'refs/remotes/origin/take-n-from-new-master' into 1.7

domsj committed:
Merge remote-tracking branch 'refs/remotes/origin/start-i-after-catchup' into 1.7

Conflicts:
	src/node/catchup.ml
	src/paxos/slave.ml

NicolasT committed:
Cleanup: Replace tabs with spaces

This commit replaces all tabs in *.ml[i] files with 2 spaces using

    sed -e 's/\t/  /g'

NicolasT committed:
Cleanup: Remove trailing whitespace

This commit removes all trailing whitespace in *.ml[i] files using

    sed -e's/[[:space:]]*$//'

NicolasT committed:
Cleanup: Remove trailing empty lines

This commit removes trailing empty lines from *.ml[i] files using

    sed -e :a -e '/^\n*$/{$d;N;ba' -e '}'

NicolasT committed:
Add ocp-indent settings file

NicolasT committed:
Cleanup: Run ocp-indent

Run `ocp-indent -i` it on *.ml[i] files.

NicolasT committed:
Merge pull request #146 from Incubaid/indent

Cleanup: Batch whitespace fixes & ocp-indent run

domsj committed:
shorten drop_master timeout to expose more sub awesome behaviour

domsj committed:
more drop_master in tests

domsj committed:
insert ElectionTimeout before going to Accepteds_check_done
refactoring in Master_consensus to reuse an existing code path

domsj committed:
forced masters should not respond to clients when they're not the master,
as this could mean some updates are not yet applied to the store

domsj committed:
change election_timeout strategy to avoid an explosion of ignored threads

domsj committed:
respect another node running for master by temporarily not sending promises to other nodes

Commit:
better logging for ARAKOON-422 (+ test)

domsj committed:
Merge remote-tracking branch 'refs/remotes/origin/timeout-accepteds-check-done' into origin-1.7

domsj committed:
Merge remote-tracking branch 'refs/remotes/origin/forced-master-need-election-too' into origin-1.7

NicolasT committed:
Add 'indent' target to Makefile

domsj committed:
change fsm thread shutdown mechanism to make it more reliable

NicolasT committed:
Merge pull request #153 from Incubaid/indent

Add 'indent' target to Makefile

Commit:
fixup.

domsj committed:
make catchup stoppable after each paxos value it has applied to the store

Commit:
fix indentation

domsj committed:
Multi_paxos.Full product_wanted should really mean full

domsj committed:
always start a lease expiration thread when entering Slave_steady_state

domsj committed:
keep lease expirations flowing in learner nodes

domsj committed:
Multi_paxos.Full product_wanted should really mean full

domsj committed:
ensure no Prepares from the same node get queued up in the respect mechanism

domsj committed:
Merge #151 remote-tracking branch 'refs/remotes/origin/respect-running-for-master' into origin-1.7

Conflicts:
	src/paxos/multi_paxos.ml
	src/paxos/multi_paxos_test.ml

domsj committed:
Merge branch '1.7' of https://github.com/Incubaid/arakoon into origin-1.7

# Please enter a commit message to explain why this merge is necessary,
# especially if it merges an updated upstream into a topic branch.
#
# Lines starting with '#' will be ignored, and an empty message aborts
# the commit.

toolslive committed:
Merge pull request #150 from Incubaid/ARAKOON-422

better logging for ARAKOON-422 (+ test)

Commit:
bollocks... %002 -> 002

NicolasT committed:
Include 'lwt.ssl' dependency

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

NicolasT committed:
CI: Install libssl-dev & OCaml 'ssl' package

domsj committed:
Merge #155 remote-tracking branch 'refs/remotes/origin/cancel-fsm' into origin-1.7

Conflicts:
	src/node/catchup.ml

domsj committed:
prevent mismatch between number of updates and update_results returned from store when inserting a value

domsj committed:
fix inject LeaseExpired event in master_consensus

domsj committed:
clear any self master sets applied to the store after initial catchup at node start

domsj committed:
for clarity log should always be the first side effect

domsj committed:
remove unneeded Value.t option argument for slave_wait_for_accept state

Commit:
updated crtl/nodes;
removed _init.ml and made a .ocamlinit

domsj committed:
remove slave_wait_for_accept state

domsj committed:
Merge #154 remote-tracking branch 'refs/remotes/origin/fix-test-expect-progress-fixed-master' into origin-1.7

domsj committed:
Merge #157 remote-tracking branch 'refs/remotes/origin/log-first-side-effect' into origin-1.7

domsj committed:
fix a bug where an already learned value is accepted again

domsj committed:
fix pair_perfect test

domsj committed:
stop the fsms in the tests so tests are better separated

domsj committed:
leftover cleanup

domsj committed:
Merge #160 remote-tracking branch 'refs/remotes/origin/slave-accept-previous' into origin-1.7

Conflicts:
	src/paxos/slave.ml

domsj committed:
Merge pull request #161 from Incubaid/tests-stop-fsm

stop the fsms in the tests so tests are better separated

domsj committed:
allow nodes to start as forced_slave (using quiesced mechanism)

domsj committed:
slave should accept previous with higher n

domsj committed:
prevent mismatch between number of updates and update_results returned from store when inserting a value

domsj committed:
prepare value in epilogue in same way as during main part of catchup updated a few comments in slave_steady_state

domsj committed:
correctly obey accepts from other nodes even when in master state

Commit:
for range queries, don't transform the array into a list
(also some preliminary work on cllio )

domsj committed:
remove unneeded value from stable_master state

domsj committed:
Merge pull request #167 from Incubaid/obey-accept

correctly obey accepts from other nodes even when in master state

domsj committed:
Merge pull request #163 from Incubaid/fix-mismatch-1.7

prevent mismatch between number of updates and update_results returned from store when inserting a value

domsj committed:
allow configuring forced_slave from the extensions

domsj committed:
use forced_slave in a bunch of tests

Commit:
micro benchmark for ranges

Commit:
removed superfluous log statements

domsj committed:
fix compression tests by optionally waiting for compression to finish when closing a tlogcollection

domsj committed:
Lwt_buffer.wait_until_empty should really wait until empty

domsj committed:
cleanup election_suggest and read_only state

domsj committed:
swallow canceled exception from fsm thread

domsj committed:
fix test_restart_single_slave_short

NicolasT committed:
Add mkcerts.sh

Note this utility script should not be used to generate certificates
etc. for a real production setup. It should be considered insecure, has
not been audited,...

domsj committed:
Merge pull request #171 from Incubaid/prevent-hanging-on-shutdown

swallow canceled exception from fsm thread

domsj committed:
Merge pull request #170 from Incubaid/fix-test-compression

Fix test compression

domsj committed:
Merge pull request #169 from Incubaid/cleanup-election-suggest

cleanup election_suggest and read_only state

domsj committed:
check if forced slave before doing quiesced operations
assert db_optimize and db_defrag throw on forced slaves
fix some tests

domsj committed:
add test to assert forced_slave can collapse

NicolasT committed:
Add tls_cert, tls_key and tls_ca_cert configuration options

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

NicolasT committed:
Add TLS support to __open_connection

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

Commit:
benchmark supports key size as well

NicolasT committed:
Implement node-to-node TLS connectivity

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

domsj committed:
test inject_as_head on forced slaves

domsj committed:
each node should have its own range of n

domsj committed:
enable all tests in right suite
remove duplicate test_prefered_master

domsj committed:
don't wrap defrag call in try_quiesced as this results in an invalid operation

Commit:
propagate arrays iso lists for other range queries as well

domsj committed:
fix test_is_progress_possible, add test_is_progress_possible_forced_slave

Commit:
extra benchmark phase for range_entries

Commit:
better layout

NicolasT committed:
Add basic TLS validation test

domsj committed:
Merge pull request #174 from Incubaid/each-node-its-own-n

each node should have its own range of n

domsj committed:
Merge #173 remote-tracking branch 'refs/remotes/origin/nose-tests-2' into origin-1.7

domsj committed:
Merge #175 remote-tracking branch 'refs/remotes/origin/ARAKOON-359-2' into origin-1.7

Conflicts:
	src/node/node_main.ml

domsj committed:
fix StoreCounterTooLow for forced slaves

toolslive committed:
Merge pull request #177 from Incubaid/fix-storecountertoolow-forced-slave

fix StoreCounterTooLow for forced slaves

toolslive committed:
Merge pull request #176 from Incubaid/range_performance

Range performance

NicolasT committed:
Add client service TLS support

See: ARAKOON-428
See: http://jira.incubaid.com/browse/ARAKOON-428

NicolasT committed:
Accept TLS connections without certificate

Certificate checking should be performed at some other level (e.g.
by using `Ssl.Verify_fail_if_no_peer_cert`).

NicolasT committed:
Implement TLS client support for CLI actions

Commit:
adding lcnum and ncnum parameters do config
WIP: we probably want to make these optional with defaults ?(lcnum=...)

Commit:
stopNode -> stopOne

toolslive committed:
Merge pull request #158 from Incubaid/ctrl_2

updated crtl/nodes;

Commit:
seed update

NicolasT committed:
Doc: Add TLS setup instructions

NicolasT committed:
Tests: Use client TLS connections in test_basic_tls

NicolasT committed:
Merge ARAKOON-428 into 1.7 (#172)

Commit:
remove comment

Commit:
Merge branch '1.7' into tokyo_cache_parameters

Conflicts:
	src/node/node_cfg.ml
	src/system/startup.ml

toolslive committed:
Merge pull request #178 from Incubaid/tokyo_cache_parameters

adding lcnum and ncnum parameters do config

Commit:
moved to shiny new camltc range_entries

Commit:
lower benchmark size for test_272

NicolasT committed:
Merge remote-tracking branch 'origin/1.6' into 1.7

* origin/1.6:
  PyLabs: Don't ignore 'inject-as-head' exit code
  1.6.3 in .deb & META
  prevent mismatch between number of updates and update_results returned from store when inserting a value

Conflicts:
	extension/test/server/shaky/test_inject_as_head.py

NicolasT committed:
Update META to 1.7+dev

NicolasT committed:
List 'ssl' as a dependency

toolslive committed:
Merge pull request #180 from Incubaid/ARAKOON-434

Arakoon 434

Commit:
python needs a compiler

Commit:
renamed Version to Arakoon_version to avoid conflicts

NicolasT committed:
Add TLS support to Python client

See: ARAKOON-433
See: http://jira.incubaid.com/browse/ARAKOON-433
See: ARAKOON-428

NicolasT committed:
CI: Add basic Python TLS client test

See: ARAKOON-433

NicolasT committed:
Merge pull request #181 from Incubaid/ARAKOON-433

TLS support for Python client

NicolasT committed:
Pylabs: Add TLS configuration methods

See: ARAKOON-432
See: http://jira.incubaid.com/browse/ARAKOON-432

NicolasT committed:
CI: Use 'TEST_SPEC' environment variable

toolslive committed:
Merge pull request #182 from Incubaid/ci-pattern

CI: Use 'TEST_SPEC' environment variable

Commit:
getting ready for ocaml 4.01

Commit:
setup_env.py now uses ocaml 4.01

Commit:
opam switch 4.01.0

toolslive committed:
Merge pull request #184 from Incubaid/ocaml_4.01_ready

Ocaml 4.01 ready

Commit:
trying out a new seed

toolslive committed:
Merge pull request #185 from Incubaid/seed_fiddling

trying out a new seed

Commit:
use a seed based on ocaml 4.01.0

Commit:
seed => 0913

toolslive committed:
Merge pull request #186 from Incubaid/seed_fiddling2

Seed fiddling2

domsj committed:
more finegrained start of lease expirations (int64 -> float)

domsj committed:
remove duplicate start of lease expiration thread

domsj committed:
start running for election or slave_fake_prepare when entering simple paxos

domsj committed:
log on info level if on_accept takes more than 1 second

domsj committed:
fix memory leak in slaves accumulating more and more leaseExpired messages and threads

domsj committed:
make sure lease expirations don't fire before the lease is expired

toolslive committed:
Merge pull request #188 from Incubaid/bump-log-level

log on info level if on_accept takes more than 1 second

Commit:
replaced shell based tls tests with python stubs

Commit:
test_ssl don't propagate exception in teardown

Commit:
don't use my paths...

toolslive committed:
Merge pull request #190 from Incubaid/system_tests_ssl

System tests ssl

toolslive committed:
Merge pull request #189 from Incubaid/unix-time-2

Unix time 2

domsj committed:
start lease expiration thread after consensus for a master has been reached
refactoring of what master_option means

Commit:
python client's multiGet now has correct order

toolslive committed:
Merge pull request #191 from Incubaid/ARAKOON-436

python client's multiGet now has correct order

domsj committed:
only sleep as much as needed for the lease to expire

domsj committed:
don't hold on to prepares received while respecting another potential master

domsj committed:
comment out erroneous assert from test_restart_master_long

toolslive committed:
Merge pull request #192 from Incubaid/fix-restart-master-long

comment out erroneous assert from test_restart_master_long

Commit:
test should work on jenkins again

toolslive committed:
Merge pull request #193 from Incubaid/lease-expiration-start

Lease expiration start

domsj committed:
Merge pull request #194 from Incubaid/disk_full

test should work on jenkins again

domsj committed:
correctly ignore accepts in slave_steady_state

domsj committed:
bump some catchup, tlog compression and file related log levels

domsj committed:
bump log level of administrative and nursery related commands

toolslive committed:
Merge pull request #195 from Incubaid/ARAKOON-439

Arakoon 439

domsj committed:
increasing counter in election_suggest loop which decreases respect for preferred masters

toolslive committed:
Merge pull request #196 from Incubaid/ignore-accepts-correctly

correctly ignore accepts in slave_steady_state

toolslive committed:
Merge pull request #197 from Incubaid/fix-test-preferred-master

increasing counter in election_suggest loop which decreases respect for ...

domsj committed:
improved logging related to lease expired events

toolslive committed:
Merge pull request #199 from Incubaid/ignore-lease-expiration-logging

improved logging related to lease expired events

domsj committed:
witness Accept messages from master in slave_steady_state

toolslive committed:
Merge pull request #200 from Incubaid/slave-on-witness

witness Accept messages from master in slave_steady_state

domsj committed:
ensure lease expirations never fire too early

domsj committed:
only hand of mastership to a preferred master if he's alive

toolslive committed:
Merge pull request #201 from Incubaid/lease-expiration-early

ensure lease expirations never fire too early

Commit:
prefix,range calls have default 1000 iso -1

a glitch in the matrix?

domsj committed:
add liveness detection loop so slaves have a failure detector too

toolslive committed:
Merge pull request #204 from Incubaid/sane_defaults

prefix,range calls have default 1000 iso -1

Commit:
2MB iso 8

Commit:
don't dump store on full ramdisk
more info logging

toolslive committed:
Merge pull request #206 from Incubaid/compress_smaller

2MB iso 8

toolslive committed:
Merge pull request #207 from Incubaid/jenkins_fix

don't dump store on full ramdisk

NicolasT committed:
Merge pull request #183 from Incubaid/ARAKOON-432

Pylabs: Add TLS configuration methods

domsj committed:
Merge pull request #205 from Incubaid/failure-detect-preferred-master

Failure detect preferred master

domsj committed:
lift liveness detection period

domsj committed:
implement --force option for --inject-as-head to allow replacing a current corrupt head.db
add different exit codes depending on which database is corrupted

domsj committed:
ARAKOON-418 fsync=true should be the default, warning otherwise

domsj committed:
disable fsync by default for most tests

NicolasT committed:
Travis: Use Anil's PPA

See: http://anil.recoil.org/2013/09/30/travis-and-ocaml.html

NicolasT committed:
Travis: Execute 'arakoon.native --run-all-tests'

Since we no longer need to build OCaml from source, we might have time
to run some real tests now within a TravisCI job slot.

Commit:
reversing the wire order (again)

NicolasT committed:
Merge pull request #211 from Incubaid/travis-ppa

Use a PPA iso building OCaml & opam from source

domsj committed:
more fsync disabled in tests; test progress possible with fsync enabled

domsj committed:
fuse the tlog catchup to allow clean shutdown during catchup

domsj committed:
rename forced_slave concept to witness node

domsj committed:
Merge pull request #212 from Incubaid/fuse-tlog-catchup

fuse the tlog catchup to allow clean shutdown during catchup

domsj committed:
Merge pull request #213 from Incubaid/witness

rename forced_slave concept to witness node

domsj committed:
fix test_fsync

toolslive committed:
Merge pull request #210 from Incubaid/ARAKOON-436_2

reversing the wire order (again)

domsj committed:
don't fsync tlog during catchup

Conflicts:
	src/node/catchup.ml

domsj committed:
fix test_learner by disabling fsync (makes the timing assumption true again)

domsj committed:
Merge pull request #208 from Incubaid/inject-as-head

implement --force option for --inject-as-head to allow replacing a curre...

domsj committed:
Merge pull request #209 from Incubaid/ARAKOON-418

ARAKOON-418 fsync=true should be the default, warning otherwise

domsj committed:
prevent lease expiration explosion for a stable master

domsj committed:
Merge pull request #215 from Incubaid/prevent-lease-expiration-explosion

prevent lease expiration explosion for a stable master

Commit:
improvements on timing,logging, defaults

toolslive committed:
Merge pull request #216 from Incubaid/benchmark_tweaking

improvements on timing,logging, defaults

domsj commented:
I'll reopen this PR against 1.7 after it passed CI

domsj committed:
copy store through a temporary file + fsync destination directory entry
